File: 3Dengine/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <map>

using namespace std;

// ========== Параметры экрана и мира ==========
const int SCREEN_WIDTH  = 120;
const int SCREEN_HEIGHT = 40;
const float PI          = 3.14159265358979f;
const float FOV         = PI / 3.0f;          // горизонтальный угол обзора
const float VERTICAL_FOV = PI / 3.0f;         // вертикальный угол обзора
const float EYE_HEIGHT   = 0.5f;              // высота камеры над полом
const float CEILING_HEIGHT = 1.0f;            // высота потолка
const float GRAVITY       = 15.0f;            // ускорение свободного падения
const float PLAYER_SPEED  = 3.0f;
const float PLAYER_ROT_SPEED = 2.0f;
const float PROJECTILE_SPEED = 10.0f;
const float LIGHT_FALLOFF_WALL   = 0.15f;
const float LIGHT_FALLOFF_FLOOR  = 0.1f;

// ========== Вспомогательные структуры ==========
struct Vec2 { float x, y; };
struct Vec3 { float r, g, b; };

// ========== ANSI‑управление консолью ==========
void clearScreen() {
    cout << "\033[2J\033[H";
}
void gotoxy(int x, int y) {
    cout << "\033[" << y << ";" << x << "H";
}
void setBackground(int r, int g, int b) {
    // clamp
    if (r < 0) r = 0; if (r > 255) r = 255;
    if (g < 0) g = 0; if (g > 255) g = 255;
    if (b < 0) b = 0; if (b > 255) b = 255;
    cout << "\033[48;2;" << r << ";" << g << ";" << b << "m";
}
void setForeground(int r, int g, int b) {
    if (r < 0) r = 0; if (r > 255) r = 255;
    if (g < 0) g = 0; if (g > 255) g = 255;
    if (b < 0) b = 0; if (b > 255) b = 255;
    cout << "\033[38;2;" << r << ";" << g << ";" << b << "m";
}
void resetColor() {
    cout << "\033[0m";
}

// ========== Типы стен ==========
struct WallType {
    Vec3 color;
};
map<char, WallType> wallTypes;

void initWallTypes() {
    wallTypes['#'] = { { 128, 128, 128 } };   // серая стена
    wallTypes['D'] = { { 139, 69, 19 } };     // коричневая
    wallTypes['='] = { { 100, 100, 100 } };   // ещё одна
}

// ========== Карта уровня ==========
class Map {
public:
    vector<string> grid;
    int width, height;

    Map(const string& filename) {
        ifstream file(filename);
        string line;
        while (getline(file, line)) {
            grid.push_back(line);
        }
        height = (int)grid.size();
        width  = (height > 0) ? (int)grid[0].length() : 0;
    }

    char getCell(int x, int y) const {
        if (x < 0 || x >= width || y < 0 || y >= height) return '#';
        return grid[y][x];
    }

    bool isWall(int x, int y) const {
        char c = getCell(x, y);
        return wallTypes.find(c) != wallTypes.end();
    }

    WallType getWallType(int x, int y) const {
        char c = getCell(x, y);
        auto it = wallTypes.find(c);
        if (it != wallTypes.end()) return it->second;
        return { {200,200,200} };
    }
};

// ========== Трассировка луча (DDA) ==========
struct HitInfo {
    float perpDist;
    int mapX, mapY;
    int side;          // 0 – вертикальная грань, 1 – горизонтальная
    WallType wall;
};

HitInfo castRay(const Map& map, float playerX, float playerY, float rayAngle) {
    float dirX = cos(rayAngle);
    float dirY = sin(rayAngle);

    int mapX = int(playerX);
    int mapY = int(playerY);

    float deltaDistX = fabs(1.0f / dirX);
    float deltaDistY = fabs(1.0f / dirY);

    float sideDistX, sideDistY;
    int stepX, stepY;

    if (dirX < 0) { stepX = -1; sideDistX = (playerX - mapX) * deltaDistX; }
    else          { stepX =  1; sideDistX = (mapX + 1.0f - playerX) * deltaDistX; }

    if (dirY < 0) { stepY = -1; sideDistY = (playerY - mapY) * deltaDistY; }
    else          { stepY =  1; sideDistY = (mapY + 1.0f - playerY) * deltaDistY; }

    bool hit = false;
    int side;
    while (!hit) {
        if (sideDistX < sideDistY) {
            sideDistX += deltaDistX;
            mapX += stepX;
            side = 0;
        } else {
            sideDistY += deltaDistY;
            mapY += stepY;
            side = 1;
        }
        if (map.isWall(mapX, mapY)) hit = true;
    }

    float perpDist;
    if (side == 0)
        perpDist = (mapX - playerX + (1.0f - stepX) / 2.0f) / dirX;
    else
        perpDist = (mapY - playerY + (1.0f - stepY) / 2.0f) / dirY;

    HitInfo info;
    info.perpDist = perpDist;
    info.mapX = mapX;
    info.mapY = mapY;
    info.side = side;
    info.wall = map.getWallType(mapX, mapY);
    return info;
}

// ========== Таблица тангенсов для пола/потолка ==========
float tanTable[SCREEN_HEIGHT];

void initTanTable() {
    float halfH = SCREEN_HEIGHT / 2.0f;
    for (int y = 0; y < SCREEN_HEIGHT; y++) {
        float angle = ((y - halfH) / halfH) * (VERTICAL_FOV / 2.0f);
        tanTable[y] = tan(angle);
    }
}

// ========== Освещение (фонарик) ==========
float lightFactorWall(float dist) {
    if (dist <= 0.0f) return 1.0f;
    float f = 1.0f / (1.0f + dist * LIGHT_FALLOFF_WALL);
    if (f < 0.0f) f = 0.0f;
    if (f > 1.0f) f = 1.0f;
    return f;
}
float lightFactorFloor(float dist) {
    if (dist <= 0.0f) return 1.0f;
    float f = 1.0f / (1.0f + dist * LIGHT_FALLOFF_FLOOR);
    if (f < 0.0f) f = 0.0f;
    if (f > 1.0f) f = 1.0f;
    return f;
}

// ========== Объекты (базовый класс) ==========
class Object {
public:
    float x, y, z;        // z – высота над полом
    float vx, vy, vz;
    float radius;
    char spriteChar;
    Vec3 color;
    int health;
    bool active;

    Object(float x_, float y_, char ch, Vec3 col, float rad = 0.3f)
        : x(x_), y(y_), z(0.0f), vx(0), vy(0), vz(0),
          radius(rad), spriteChar(ch), color(col), health(1), active(true) {}

    virtual ~Object() {}

    virtual void Update(float dt) {
        vz -= GRAVITY * dt;
        x += vx * dt;
        y += vy * dt;
        z += vz * dt;
        if (z < 0.0f) { z = 0.0f; vz = 0.0f; }
    }
};

// ========== Игрок (управляется с клавиатуры) ==========
class Player : public Object {
public:
    float angle;

    Player(float x_, float y_, float ang_)
        : Object(x_, y_, 'P', {255,255,0}, 0.2f), angle(ang_) {
        z = 0.0f; // камера всегда на полу
    }

    void Update(float dt) override {
        // Гравитация на игрока не действует
        x += vx * dt;
        y += vy * dt;
        // wall collision разрешается отдельно в main
    }
};

// ========== Снаряд ==========
class Projectile : public Object {
public:
    Projectile(float x_, float y_, float z_, float direction)
        : Object(x_, y_, '*', {255,255,0}, 0.1f) {
        z = z_;
        vx = cos(direction) * PROJECTILE_SPEED;
        vy = sin(direction) * PROJECTILE_SPEED;
        vz = 0.0f;
    }

    void Update(float dt) override {
        Object::Update(dt);
        // снаряд будет удалён при столкновении со стеной в main
    }
};

// ========== Главный цикл ==========
int main() {
    // ---------- инициализация ----------
    initWallTypes();
    initTanTable();
    Map map("map.txt");   // файл с картой должен существовать

    Player player(2.5f, 2.5f, 0.0f);
    vector<Object*> objects;

    // несколько статических объектов
    objects.push_back(new Object(4.0f, 3.0f, 'M', {255,0,0}, 0.3f));
    objects.push_back(new Object(6.0f, 5.0f, 'B', {0,255,0}, 0.3f));

    float dt = 0.05f;       // фиксированный шаг на кадр
    float focalLen = (SCREEN_HEIGHT / 2.0f) / tan(VERTICAL_FOV / 2.0f);
    float zBuffer[SCREEN_WIDTH];
    bool running = true;

    // ---------- игровой цикл ----------
    while (running) {
        // 1. Очистка и заливка пола / потолка
        clearScreen();
        for (int y = 0; y < SCREEN_HEIGHT; y++) {
            int halfH = SCREEN_HEIGHT / 2;
            float rowDist;
            Vec3 baseColor;
            if (y > halfH) {   // пол
                float tanVal = tanTable[y];
                if (fabs(tanVal) < 0.0001f) tanVal = 0.0001f;
                rowDist = EYE_HEIGHT / tanVal;
                baseColor = { 80, 80, 80 };
            } else {           // потолок
                float tanVal = -tanTable[y];
                if (tanVal < 0.0001f) tanVal = 0.0001f;
                rowDist = (CEILING_HEIGHT - EYE_HEIGHT) / tanVal;
                baseColor = { 60, 60, 100 };
            }
            float light = lightFactorFloor(rowDist);
            int r = int(baseColor.r * light);
            int g = int(baseColor.g * light);
            int b = int(baseColor.b * light);
            setBackground(r, g, b);
            gotoxy(1, y + 1);
            for (int i = 0; i < SCREEN_WIDTH; i++) cout << ' ';
        }
        resetColor();

        // 2. Стены
        for (int x = 0; x < SCREEN_WIDTH; x++) {
            float rayAngle = player.angle - FOV/2.0f +
                             (x / (float)SCREEN_WIDTH) * FOV;
            HitInfo hit = castRay(map, player.x, player.y, rayAngle);
            zBuffer[x] = hit.perpDist;

            int lineHeight = (int)(SCREEN_HEIGHT / hit.perpDist);
            int wallTop    = max(0, (SCREEN_HEIGHT - lineHeight) / 2);
            int wallBottom = min(SCREEN_HEIGHT - 1, wallTop + lineHeight - 1);

            float light = lightFactorWall(hit.perpDist);
            Vec3 col = hit.wall.color;
            int r = int(col.r * light);
            int g = int(col.g * light);
            int b = int(col.b * light);

            for (int y = wallTop; y <= wallBottom; y++) {
                gotoxy(x + 1, y + 1);
                setBackground(r, g, b);
                cout << ' ';
            }
        }
        resetColor();

        // 3. Спрайты (объекты)
        sort(objects.begin(), objects.end(),
             [&](Object* a, Object* b) {
                 float da = (a->x - player.x)*(a->x - player.x) +
                            (a->y - player.y)*(a->y - player.y);
                 float db = (b->x - player.x)*(b->x - player.x) +
                            (b->y - player.y)*(b->y - player.y);
                 return da > db;   // дальние — первыми
             });

        for (auto obj : objects) {
            if (!obj->active) continue;
            float dx = obj->x - player.x;
            float dy = obj->y - player.y;
            float dist = sqrt(dx*dx + dy*dy);

            float spriteAngle = atan2(dy, dx) - player.angle;
            while (spriteAngle >  PI) spriteAngle -= 2.0f * PI;
            while (spriteAngle < -PI) spriteAngle += 2.0f * PI;

            if (fabs(spriteAngle) > FOV/2.0f + 0.2f) continue;

            float screenX = (SCREEN_WIDTH / 2.0f) *
                            (1.0f + tan(spriteAngle) / tan(FOV/2.0f));
            int spriteScreenX = int(screenX);
            if (spriteScreenX < 0 || spriteScreenX >= SCREEN_WIDTH) continue;

            float invDist = 1.0f / dist;
            float bottomY = SCREEN_HEIGHT/2.0f -
                            (EYE_HEIGHT - obj->z) * invDist * focalLen;
            float topY    = SCREEN_HEIGHT/2.0f -
                            (EYE_HEIGHT - (obj->z + 1.0f)) * invDist * focalLen;
            int drawStartY = max(0, (int)(topY));
            int drawEndY   = min(SCREEN_HEIGHT - 1, (int)(bottomY));

            float light = lightFactorWall(dist);
            int r = int(obj->color.r * light);
            int g = int(obj->color.g * light);
            int b = int(obj->color.b * light);

            for (int y = drawStartY; y <= drawEndY; y++) {
                if (dist < zBuffer[spriteScreenX]) {
                    gotoxy(spriteScreenX + 1, y + 1);
                    setBackground(r, g, b);
                    setForeground(0, 0, 0);
                    cout << obj->spriteChar;
                }
            }
        }
        resetColor();

        // 4. Ввод команды
        gotoxy(1, SCREEN_HEIGHT + 1);
        cout << "Command (w/a/s/d/q/e/space/x): ";
        char cmd;
        cin >> cmd;

        // 5. Обработка движения игрока
        float moveSpeed = PLAYER_SPEED * dt;
        float rotSpeed  = PLAYER_ROT_SPEED * dt;
        float moveX = 0.0f, moveY = 0.0f;
        switch (cmd) {
            case 'w': moveX =  cos(player.angle) * moveSpeed;
                      moveY =  sin(player.angle) * moveSpeed; break;
            case 's': moveX = -cos(player.angle) * moveSpeed;
                      moveY = -sin(player.angle) * moveSpeed; break;
            case 'a': moveX =  cos(player.angle - PI/2.0f) * moveSpeed;
                      moveY =  sin(player.angle - PI/2.0f) * moveSpeed; break;
            case 'd': moveX =  cos(player.angle + PI/2.0f) * moveSpeed;
                      moveY =  sin(player.angle + PI/2.0f) * moveSpeed; break;
            case 'q': player.angle -= rotSpeed; break;
            case 'e': player.angle += rotSpeed; break;
            case ' ': {
                Projectile* proj = new Projectile(player.x, player.y,
                                                  player.z + 0.5f, player.angle);
                objects.push_back(proj);
                break;
            }
            case 'x': running = false; break;
        }

        // коллизия игрока со стенами (покадровое смещение)
        float newX = player.x + moveX;
        float newY = player.y + moveY;
        if (!map.isWall(int(newX), int(player.y))) player.x = newX;
        if (!map.isWall(int(player.x), int(newY))) player.y = newY;

        // 6. Обновление объектов
        for (auto obj : objects) {
            if (!obj->active) continue;
            // сохраняем старые координаты на случай отката
            float oldX = obj->x, oldY = obj->y;
            obj->Update(dt);

            // простая коллизия со стенами – если попали в стену, возвращаем объект
            if (map.isWall(int(obj->x), int(obj->y))) {
                obj->x = oldX;
                obj->y = oldY;
                // для снарядов – деактивируем при касании стены
                if (dynamic_cast<Projectile*>(obj)) {
                    obj->active = false;
                }
            }

            // удаляем объекты, улетевшие далеко или с нулевым здоровьем
            if (obj->health <= 0) obj->active = false;
        }

        // удаляем неактивные объекты
        objects.erase(
            remove_if(objects.begin(), objects.end(),
                      [](Object* o) { return !o->active; }),
            objects.end());
    }

    // очистка памяти
    for (auto o : objects) delete o;
    return 0;
}

File: 3Dengine/main.py
import sys
import math
import time
import threading
import queue

# --------------------- constants ---------------------
SCREEN_WIDTH = 120
SCREEN_HEIGHT = 40
FOV = math.pi / 3               # 60 degrees
HALF_FOV = FOV / 2
MAX_DEPTH = 25.0
GRAVITY = -9.81                 # units/s² down
FRAME_DT = 1 / 30               # desired frame time
PLAYER_EYE_HEIGHT = 0.6         # eye above ground

# shading: distance -> (character, ansi-colour)
SHADE_LEVELS = [
    (2.0,  '█', '97'),   # bright white
    (5.0,  '▓', '37'),   # white
    (8.0,  '▒', '90'),   # bright black (gray)
    (12.0, '░', '90'),
    (16.0, '.', '90'),
    (20.0, ' ', '0')     # invisible
]
# ansi colour helpers
def ansi_fg(col):
    return f'\033[38;5;{col}m'
def ansi_reset():
    return '\033[0m'

def shade_for_dist(dist):
    """return (char, ansi_fg_code) for given perpendicular distance"""
    for max_d, ch, col in SHADE_LEVELS:
        if dist <= max_d:
            return ch, col
    return ' ', '0'

# --------------------- map ---------------------
class Map:
    def __init__(self, filename):
        with open(filename, 'r') as f:
            lines = [line.rstrip('\n') for line in f if line.strip() != '']
        self.height = len(lines)
        self.width = max(len(line) for line in lines)
        self.grid = []
        self.floor_z = []      # floor height of each cell
        self.ceiling_z = []    # ceiling height
        self.stairs_up = []    # bool
        self.stairs_down = []
        for y in range(self.height):
            row = []
            fz_row = []
            cz_row = []
            su_row = []
            sd_row = []
            for x in range(self.width):
                ch = lines[y][x] if x < len(lines[y]) else ' '
                row.append(ch)
                # default floor 0, ceiling 2
                fz = 0.0
                cz = 2.0
                su = False
                sd = False
                if ch == '#' or ch == 'D':   # solid wall
                    fz = 0.0
                    cz = 2.0
                elif ch == '^':               # stairs up
                    su = True
                    fz = 1.0                  # raised floor
                    cz = 3.0
                elif ch == 'v':               # stairs down
                    sd = True
                    fz = -1.0
                    cz = 1.0
                # other: floor, keep default
                fz_row.append(fz)
                cz_row.append(cz)
                su_row.append(su)
                sd_row.append(sd)
            self.grid.append(row)
            self.floor_z.append(fz_row)
            self.ceiling_z.append(cz_row)
            self.stairs_up.append(su_row)
            self.stairs_down.append(sd_row)

    def is_wall(self, x, y):
        if x < 0 or y < 0 or x >= self.width or y >= self.height:
            return True
        ch = self.grid[y][x]
        return ch in ('#', 'D')

    def wall_height(self, x, y):
        """height of wall block (top - bottom)"""
        if self.is_wall(x, y):
            return self.ceiling_z[y][x] - self.floor_z[y][x]
        return 0.0

    def floor_at(self, x, y):
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.floor_z[y][x]
        return 0.0

    def is_stairs_up(self, x, y):
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.stairs_up[y][x]
        return False

    def is_stairs_down(self, x, y):
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.stairs_down[y][x]
        return False

# --------------------- vector math ---------------------
def rotate_vec(x, y, angle):
    c = math.cos(angle)
    s = math.sin(angle)
    return x * c - y * s, x * s + y * c

# --------------------- game object ---------------------
class GameObject:
    def __init__(self, x, y, z, char, color, health=1):
        self.x = x
        self.y = y
        self.z = z
        self.vx = 0.0
        self.vy = 0.0
        self.vz = 0.0
        self.char = char
        self.color = color
        self.health = health
        self.alive = True
        self.radius = 0.3          # for collision / interaction

    def update(self, dt, game_map, all_objects):
        """physics + interactions. returns list of events"""
        # gravity
        self.vz += GRAVITY * dt
        # movement
        new_x = self.x + self.vx * dt
        new_y = self.y + self.vy * dt
        new_z = self.z + self.vz * dt

        # simple collision with walls (2D)
        if not game_map.is_wall(int(new_x), int(self.y)):
            self.x = new_x
        if not game_map.is_wall(int(self.x), int(new_y)):
            self.y = new_y

        # vertical collision: floor
        floor_h = game_map.floor_at(int(self.x), int(self.y))
        if new_z < floor_h:
            self.z = floor_h
            self.vz = 0
        else:
            self.z = new_z

        # check if dead (fell off map)
        if self.z < -10:
            self.alive = False

        # stair interactions: change player z (handled in player logic)
        # here just for generic objects we don't use stairs.

        # interactions with other objects
        events = []
        for other in all_objects:
            if other is self or not other.alive:
                continue
            dx = self.x - other.x
            dy = self.y - other.y
            dz = self.z - other.z
            dist = math.sqrt(dx*dx + dy*dy + dz*dz)
            if dist < self.radius + other.radius:
                # collision: simple damage transfer (e.g., projectile vs enemy)
                if hasattr(self, 'damage') and self.damage > 0:
                    other.health -= self.damage
                    events.append(f'{other.char} takes {self.damage} damage')
                    self.alive = False   # projectile disappears
                # bounce: simple
                if self.vx or self.vy:
                    # push apart
                    pass
        return events

# --------------------- camera ---------------------
class Camera:
    def __init__(self, x, y, z, angle):
        self.x = x
        self.y = y
        self.z = z
        self.angle = angle
        self.pitch = 0.0   # looking straight ahead

# --------------------- ray casting ---------------------
def cast_ray(cam, game_map, screen_x):
    """cast a ray for column screen_x, return (perp_dist, wall_height, side)"""
    ray_angle = cam.angle + math.atan((2 * screen_x / SCREEN_WIDTH - 1) * math.tan(HALF_FOV))
    ray_dir_x = math.cos(ray_angle)
    ray_dir_y = math.sin(ray_angle)

    # DDA
    map_x = int(cam.x)
    map_y = int(cam.y)
    delta_dist_x = abs(1 / (ray_dir_x + 1e-10))
    delta_dist_y = abs(1 / (ray_dir_y + 1e-10))

    if ray_dir_x < 0:
        step_x = -1
        side_dist_x = (cam.x - map_x) * delta_dist_x
    else:
        step_x = 1
        side_dist_x = (map_x + 1.0 - cam.x) * delta_dist_x

    if ray_dir_y < 0:
        step_y = -1
        side_dist_y = (cam.y - map_y) * delta_dist_y
    else:
        step_y = 1
        side_dist_y = (map_y + 1.0 - cam.y) * delta_dist_y

    hit = False
    side = 0
    while not hit:
        if side_dist_x < side_dist_y:
            side_dist_x += delta_dist_x
            map_x += step_x
            side = 0
        else:
            side_dist_y += delta_dist_y
            map_y += step_y
            side = 1
        if game_map.is_wall(map_x, map_y):
            hit = True

    if side == 0:
        perp_dist = (map_x - cam.x + (1 - step_x) / 2) / (ray_dir_x + 1e-10)
    else:
        perp_dist = (map_y - cam.y + (1 - step_y) / 2) / (ray_dir_y + 1e-10)

    wall_h = game_map.wall_height(map_x, map_y)
    return perp_dist, wall_h, side

# --------------------- rendering ---------------------
def render_frame(cam, game_map, objects, screen_buf, z_buffer):
    """fill screen_buf (list of strings) and z_buffer with distances"""
    # clear
    for y in range(SCREEN_HEIGHT):
        screen_buf[y] = [' '] * SCREEN_WIDTH
        z_buffer[y] = [MAX_DEPTH] * SCREEN_WIDTH

    # ceiling / floor background (simple)
    for y in range(SCREEN_HEIGHT):
        if y < SCREEN_HEIGHT/2:
            char = '·'
            col = '34'   # blue
        else:
            char = '~'
            col = '32'   # green
        for x in range(SCREEN_WIDTH):
            screen_buf[y][x] = f'{ansi_fg(col)}{char}{ansi_reset()}'

    # cast walls
    for x in range(SCREEN_WIDTH):
        perp_dist, wall_h, side = cast_ray(cam, game_map, x)
        if perp_dist <= 0:
            perp_dist = 0.01
        # projection
        scale = SCREEN_HEIGHT / perp_dist
        eye_z = cam.z + PLAYER_EYE_HEIGHT
        wall_top = SCREEN_HEIGHT/2 - (wall_h - eye_z) * scale
        wall_bot = SCREEN_HEIGHT/2 - (0 - eye_z) * scale

        # shading
        shade_char, shade_col = shade_for_dist(perp_dist)
        if side == 1:   # y-side, slightly darker
            shade_col = '90' if shade_col == '37' else shade_col

        # clamp to screen
        top = max(0, int(wall_top))
        bot = min(SCREEN_HEIGHT-1, int(wall_bot))
        for y in range(top, bot+1):
            screen_buf[y][x] = f'{ansi_fg(shade_col)}{shade_char}{ansi_reset()}'
            z_buffer[y][x] = perp_dist   # store for sprite clipping

    # sprites (objects)
    sprite_list = []
    for obj in objects:
        if not obj.alive:
            continue
        # vector from camera to object
        dx = obj.x - cam.x
        dy = obj.y - cam.y
        # transform to camera space
        inv_angle = -cam.angle
        tx, ty = rotate_vec(dx, dy, inv_angle)
        if tx <= 0:
            continue   # behind
        # perspective projection
        sprite_screen_x = int((SCREEN_WIDTH/2) * (1 + ty / tx / math.tan(HALF_FOV)))
        sprite_height = int(SCREEN_HEIGHT / tx)   # object height in pixels
        dz = obj.z - cam.z
        sprite_screen_y = int(SCREEN_HEIGHT/2 - (dz - PLAYER_EYE_HEIGHT) * SCREEN_HEIGHT / tx)

        sprite_list.append((tx, sprite_screen_x, sprite_height, sprite_screen_y, obj))

    # sort far to near
    sprite_list.sort(key=lambda s: s[0], reverse=True)

    for _, sx, sh, sy, obj in sprite_list:
        # clipping
        draw_start_x = max(0, sx - sh//2)
        draw_end_x = min(SCREEN_WIDTH-1, sx + sh//2)
        draw_start_y = max(0, sy - sh//2)
        draw_end_y = min(SCREEN_HEIGHT-1, sy + sh//2)

        for x in range(draw_start_x, draw_end_x+1):
            if x < 0 or x >= SCREEN_WIDTH:
                continue
            # check depth buffer
            if z_buffer[draw_start_y][x] < obj.x - cam.x:   # approximate
                continue
            for y in range(draw_start_y, draw_end_y+1):
                if y < 0 or y >= SCREEN_HEIGHT:
                    continue
                # distance based fading
                dist = math.sqrt((obj.x-cam.x)**2 + (obj.y-cam.y)**2)
                _, col = shade_for_dist(dist)
                screen_buf[y][x] = f'{ansi_fg(col)}{obj.char}{ansi_reset()}'

def display(screen_buf):
    """print whole screen buffer with ANSI clear"""
    sys.stdout.write('\033[2J\033[H')   # clear screen, home
    for row in screen_buf:
        sys.stdout.write(''.join(row) + '\n')
    sys.stdout.flush()

# --------------------- input thread ---------------------
input_queue = queue.Queue()

def input_reader():
    while True:
        try:
            line = sys.stdin.readline()
            if not line:
                break
            cmd = line.strip().lower()
            input_queue.put(cmd)
        except:
            break

# --------------------- main game loop ---------------------
def main():
    # load map
    try:
        game_map = Map('level.txt')
    except FileNotFoundError:
        print("level.txt not found. Creating a default one.")
        with open('level.txt', 'w') as f:
            f.write("####################\n")
            f.write("#   ^    #    v   #\n")
            f.write("#   #         #   #\n")
            f.write("#        #        #\n")
            f.write("#   D    #    D   #\n")
            f.write("####################\n")
        game_map = Map('level.txt')

    # camera start position
    cam = Camera(x=1.5, y=1.5, z=0.0, angle=0.0)

    # objects: enemy and projectiles
    enemy = GameObject(5.0, 1.5, 0.0, 'E', '196', health=3)  # red-ish
    objects = [enemy]
    player_bullets = []   # separate list for projectiles

    # screen buffers
    screen_buf = [[' ']*SCREEN_WIDTH for _ in range(SCREEN_HEIGHT)]
    z_buffer = [[MAX_DEPTH]*SCREEN_WIDTH for _ in range(SCREEN_HEIGHT)]

    # start input thread
    input_thread = threading.Thread(target=input_reader, daemon=True)
    input_thread.start()

    # timings
    last_time = time.time()
    running = True

    while running:
        # process input commands
        while not input_queue.empty():
            cmd = input_queue.get_nowait()
            if cmd == 'quit':
                running = False
                break
            elif cmd == 'w':
                # move forward
                dx = math.cos(cam.angle) * 0.2
                dy = math.sin(cam.angle) * 0.2
                new_x = cam.x + dx
                new_y = cam.y + dy
                if not game_map.is_wall(int(new_x), int(new_y)):
                    cam.x = new_x
                    cam.y = new_y
            elif cmd == 's':
                dx = -math.cos(cam.angle) * 0.2
                dy = -math.sin(cam.angle) * 0.2
                new_x = cam.x + dx
                new_y = cam.y + dy
                if not game_map.is_wall(int(new_x), int(new_y)):
                    cam.x = new_x
                    cam.y = new_y
            elif cmd == 'a':
                # strafe left
                dx = -math.sin(cam.angle) * 0.2
                dy = math.cos(cam.angle) * 0.2
                new_x = cam.x + dx
                new_y = cam.y + dy
                if not game_map.is_wall(int(new_x), int(new_y)):
                    cam.x = new_x
                    cam.y = new_y
            elif cmd == 'd':
                dx = math.sin(cam.angle) * 0.2
                dy = -math.cos(cam.angle) * 0.2
                new_x = cam.x + dx
                new_y = cam.y + dy
                if not game_map.is_wall(int(new_x), int(new_y)):
                    cam.x = new_x
                    cam.y = new_y
            elif cmd == 'e':
                cam.angle += 0.1
            elif cmd == 'q':
                cam.angle -= 0.1
            elif cmd == ' ':
                # jump
                if cam.z <= game_map.floor_at(int(cam.x), int(cam.y)) + 0.01:
                    cam.vz = 3.0   # upward velocity
                    cam.z += 0.01  # unstick
            elif cmd == 'f':
                # shoot projectile from camera
                bullet = GameObject(cam.x, cam.y, cam.z + PLAYER_EYE_HEIGHT,
                                    '*', '226', health=1)
                bullet.damage = 1
                speed = 8.0
                bullet.vx = math.cos(cam.angle) * speed
                bullet.vy = math.sin(cam.angle) * speed
                bullet.vz = 0.0
                bullet.radius = 0.1
                player_bullets.append(bullet)
                objects.append(bullet)

        # timing
        current_time = time.time()
        dt = current_time - last_time
        if dt < FRAME_DT:
            time.sleep(FRAME_DT - dt)
            current_time = time.time()
            dt = current_time - last_time
        last_time = current_time

        if not running:
            break

        # update player physics (gravity)
        cam.vz += GRAVITY * dt if hasattr(cam, 'vz') else 0
        if not hasattr(cam, 'vz'):
            cam.vz = 0.0
        new_z = cam.z + cam.vz * dt
        floor_h = game_map.floor_at(int(cam.x), int(cam.y))
        if new_z < floor_h:
            cam.z = floor_h
            cam.vz = 0
        else:
            cam.z = new_z

        # check stairs
        if game_map.is_stairs_up(int(cam.x), int(cam.y)):
            cam.z += 1.0   # teleport up
        elif game_map.is_stairs_down(int(cam.x), int(cam.y)):
            cam.z -= 1.0

        # update objects
        all_objs = objects + player_bullets
        for obj in all_objs:
            if obj.alive:
                events = obj.update(dt, game_map, all_objs)
                for ev in events:
                    # could print messages later
                    pass

        # remove dead objects
        objects = [obj for obj in objects if obj.alive]
        player_bullets = [b for b in player_bullets if b.alive]

        # simple enemy AI: move toward player slowly
        if enemy.alive:
            dx = cam.x - enemy.x
            dy = cam.y - enemy.y
            dist = math.sqrt(dx*dx+dy*dy)
            if dist > 1.0:
                speed = 1.0
                enemy.vx = (dx/dist) * speed
                enemy.vy = (dy/dist) * speed
            else:
                enemy.vx = enemy.vy = 0.0

        # render
        render_frame(cam, game_map, objects, screen_buf, z_buffer)
        display(screen_buf)

    # cleanup
    sys.stdout.write('\033[2J\033[H')
    sys.stdout.write('Game exited.\n')

if __name__ == '__main__':
    main()
